docs/Custom-Protos.md

```diff
@@ -0,0 +1,167 @@
+# Creating custom protobuffer messages
+
+Unity and Python communicate by sending protobuffer messages to and from each other. You can create custom protobuffer messages if you want to exchange structured data beyond what is included by default. 
+
+Whenever you change the fields of a custom message, you must run `protobuf-definitions/make.bat` to create C# and Python files corresponding to the new message. Follow the directions in that file for guidance. After running it, reinstall the Python package by running `pip install ml-agents` and make sure your Unity project is using the newly-generated version of `UnitySDK`.
+
+## Custom message types
+
+There are three custom message types currently supported, described below. In each case, `env` is an instance of a `UnityEnvironment` in Python. `CustomAction` is described most thoroughly; usage of the  other custom messages follows a similar template. 
+
+### Custom actions
+
+By default, the Python API sends actions to Unity in the form of a floating-point list per agent and an optional string-valued text action. 
+
+You can define a custom action type to replace or augment this by adding fields to the `CustomAction` message, which you can do by editing the file `protobuf-definitions/proto/mlagents/envs/communicator_objects/custom_action.proto`. 
+
+Instances of custom actions are set via the `custom_action` parameter of `env.step`. An agent receives a custom action by defining a method with the signature
+
+```csharp
+public virtual void AgentAction(float[] vectorAction, string textAction, CommunicatorObjects.CustomAction customAction)
+```
+
+Here is an example of creating a custom action that instructs an agent to choose a cardinal direction to walk in and how far to walk. 
+
+`custom_action.proto` will look like 
+
+```protobuf
+syntax = "proto3";
+
+option csharp_namespace = "MLAgents.CommunicatorObjects";
+package communicator_objects;
+
+message CustomAction {
+    enum Direction {
+        NORTH=0;
```

Comment:
These values should be different, otherwise this proto won't compile.

Reply:
Good catch, copy/paste error from a working example.